### PR TITLE
Patterns: create patterns package and share the create pattern modal between post editor and site editor

### PR DESCRIPTION
**Pull request by glendaviesnz**
## What?
Shares a create pattern modal between the post and the site editor via a newly created `patterns` package. This new package will take over all management of synced and unsynced user created patterns to allow the eventual deprecation of the `reusable-blocks` package.

## Why?
Currently there is duplicated code for adding patterns in post editor (imported from `reusable-blocks`) and site editor (local modal component), which means any changes need to be applied in two places.

## How?
Adds a CreatePatternModel in a new `patterns` package.

## Testing Instructions

- In the post editor add a block and use the `Create pattern` block menu to create both a synced and an unsynced pattern and make sure both options work as expected
- In the post editor make sure the Detach pattern and Manage patterns options on the block overflow menu work correctly
- In the site editor pattern section use the + at the top of the left navigation to create both a synced and unsynced pattern and make sure both are created as expected
- In the site editor block editor use the `Create pattern` block menu to create both a synced and an unsynced pattern and make sure both options work as expected and also make sure the Detach pattern and Manage patterns options on the block overflow menu work correctly

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/3629020/d5d7b369-5369-48e3-801c-95d3bdd615f3

### CODE REVIEW

**Comment by glendaviesnz** [comment]
@aaronrobertshaw I initially started off with the shared model in the reusable-blocks package, but this doesn't help with the longer term aim of deprecating that package. It didn't seem like it was worth creating a `patterns` package at this point so the `editor` package seemed like the obvious place to put this, but open to feedback on whether this is the right home for it or not.

**Comment by glendaviesnz** [comment]
Has broken e2e tests - will look into that.

**Comment by kevin940726** [comment]
Looks good! Do we still need the `@wordpress/reusable-blocks` package in this case? Should we deprecate it? If we should, should we do it in this PR or future PRs?

**Comment by aaronrobertshaw** [comment]
> Do we still need the @wordpress/reusable-blocks package in this case? Should we deprecate it? If we should, should we do it in this PR or future PRs?

I believe the general thinking here was we'd look to deprecate the reusable blocks package as part of a larger follow-up that would complete the renaming of reusable blocks to patterns within code etc.

**Comment by aaronrobertshaw** [comment]
> I initially started off with the shared model in the reusable-blocks package, but this doesn't help with the longer term aim of deprecating that package. 

Agreed 👍 

> It didn't seem like it was worth creating a patterns package at this point so the editor package seemed like the obvious place to put this, but open to feedback on whether this is the right home for it or not.

I'm not sure where it best belongs. I don't see many other similar components in the `editor` package. I was probably expecting it in the `block-editor` package.

Wherever it lands, we should update the PR description to match.

**Comment by aaronrobertshaw** [comment]
Can this shared modal also be used in the `reusable-blocks package` for `ReusableBlocksMenuItems` so that it is reused for the block editor in the Site Editor and widget areas?

**Comment by Mamaduka** [comment]
FYI, I think there is or was a plan to deprecate the `editor` package and absorb WP-specific logic in the `edit-*` packages; @youknowriad mentioned this a few times recently.

**Comment by glendaviesnz** [comment]
Thanks for the feedback @kevin940726, @aaronrobertshaw, @Mamaduka 

>Looks good! Do we still need the @wordpress/reusable-blocks package in this case? Should we deprecate it? If we should, should we do it in this PR or future PRs?

As @aaronrobertshaw suggests I think we should deprecate it down the track. Some parts of it are still used, eg. the delete options in edit-site, and I think we should limit the scope of this PR to just sharing the create modal, but in a way that allows us to then follow up with moving the remaining functionality and deprecating the reusable-blocks package. To do it all in one PR will make it too hard to fully review/test I think.


>  I was probably expecting it in the block-editor package.

Because it used core-data entities it can't go in block-editor as that package is not allowed to have anything WP specific.

> Can this shared modal also be used in the reusable-blocks package for ReusableBlocksMenuItems so that it is reused for the block editor in the Site Editor and widget area

I don't think we should share it back into reusable blocks package, the idea is to locate it somewhere that will enable us to replace all existing uses of `ReusableBlocksMenuItems` with `PatternsMenuItems`.

>  I think there is or was a plan to deprecate the editor package and absorb WP-specific logic in the edit-* packages;

Thanks for this, will look into this a bit more, perhaps it might be worth adding an `edit-patterns` package as once we get into the partially synced patterns there is going to be a lot more complexity that will need to live somewhere.

**Comment by glendaviesnz** [comment]
I have moved the modal to a new `edit-patterns` package which I think is going to be a better long term solution, especially if the `editor` package is being deprecated.

File: packages/edit-patterns/src/components/create-pattern-modal.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+		[ createPattern, clientIds, onSuccess, createErrorNotice, onError ]
+	);
+	return (
+		<Modal
+			title={ __( 'Create pattern' ) }
+			onRequestClose={ () => {
+				onClose();
+				setTitle( '' );
+			} }
+			overlayClassName="patterns-menu-items__convert-modal"
```
Comment (aaronrobertshaw) [inline packages/edit-patterns/src/components/create-pattern-modal.js]: What do you think of avoiding hardcoded class names? I know this isn't the `components` package but it could provide freedom down the line

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for the explanations

> I don't think we should share it back into reusable blocks package, the idea is to locate it somewhere that will enable us to replace all existing uses of ReusableBlocksMenuItems with PatternsMenuItems.

Let me rephrase that question then. Can we not already replace `ReusableBlocksMenuItems` with `PatternsMenuItems` in the [edit-site block editor](https://github.com/WordPress/gutenberg/blob/afae324ecd3e22de30121f1cd3a233ee0f3b7229/packages/edit-site/src/components/block-editor/index.js#L37) and the [widgets area block editor](https://github.com/WordPress/gutenberg/blob/afae324ecd3e22de30121f1cd3a233ee0f3b7229/packages/edit-widgets/src/components/widget-areas-block-editor-provider/index.js#L115)?

I don't think those two cases expand the scope of this PR much at all (especially given the new edit-patterns package). They will also make sure the new shared components line up with existing use cases.


> I have moved the modal to a new edit-patterns package which I think is going to be a better long term solution, especially if the editor package is being deprecated.

The new `edit-patterns` package is testing well for me so far. I think it would be good to get some wider confirmation on the approach as well. 

Very nice work! ✨

**Comment by glendaviesnz** [comment]
> Can we not already replace ReusableBlocksMenuItems with PatternsMenuItems in the [edit-site block editor](https://github.com/WordPress/gutenberg/blob/afae324ecd3e22de30121f1cd3a233ee0f3b7229/packages/edit-site/src/components/block-editor/index.js#L37) and the [widgets area block editor](https://github.com/WordPress/gutenberg/blob/afae324ecd3e22de30121f1cd3a233ee0f3b7229/packages/edit-widgets/src/components/widget-areas-block-editor-provider/index.js#L115)?

Ah, gotcha, have done that now.

File: packages/edit-patterns/src/components/create-pattern-modal.js (hunk 1 above)
Comment (glendaviesnz) [reply packages/edit-patterns/src/components/create-pattern-modal.js]: Have added an optional className prop

**Comment by youknowriad** [comment]
just a small remark. I think `edit-*` package convention is meant for top level packages that actually call `React.render` which is probably not the case here.

File: packages/edit-patterns/src/components/index.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import { useSelect } from '@wordpress/data';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+
+/**
+ * Internal dependencies
+ */
+import PatternConvertButton from './pattern-convert-button';
+import PatternsManageButton from './patterns-manage-button';
+
+export default function PatternsMenuItems( { rootClientId } ) {
```
Comment (youknowriad) [inline packages/edit-patterns/src/components/index.js]: It's a bit weird to me that this component is basically expected to be rendered within a `BlockEditorProvider` to actually work properly. 

Do you think there's a possibility to remove the dependency to `block-editor` entirely to make the component more reusable?

If we are to keep it as it is, we should clarify that this package is not about "patterns" but about "using patterns within a block editor" or something like that.

**Comment by glendaviesnz** [comment]
> just a small remark. I think `edit-*` package convention is meant for top level packages that actually call `React.render` which is probably not the case here.

I was just going from [this comment](https://github.com/WordPress/gutenberg/pull/53161#issuecomment-1658072417). Thanks for the extra context @youknowriad, I have just renamed it to `patterns` which is closer to the `reusable-blocks` package that it is replacing.

File: packages/edit-patterns/src/components/index.js (hunk 2 above)
Comment (glendaviesnz) [reply packages/edit-patterns/src/components/index.js]: Thanks for the feedback @youknowriad. Currently, it is just a copy of how the equivalent [reusable block component](https://github.com/WordPress/gutenberg/blob/trunk/packages/reusable-blocks/src/components/reusable-blocks-menu-items/index.js#L13) that it is deprecating was set up.

It is a bit hard to see how this particular component could be reusable outside an editor context as it augments the editor block menus, and needs access to currently selected blocks, etc., which could be passed in as props instead of deps, but hard to see what they would be swapped out for in a non-editor context.

I did initially just add these components to the `editor` package, but moved it to a new package after [this comment](https://github.com/WordPress/gutenberg/pull/53161#issuecomment-1658072417). 

I am open to ideas about where these new components that deprecate the reusable block ones can best be located.

Comment (youknowriad) [reply packages/edit-patterns/src/components/index.js]: The main reason I don't like it being in `editor` is because editor is about a "post" context which is not always the case in the site editor. The site editor ideally shouldn't have a dependency to "editor". I know it's already the case because of ... other mistakes.

Comment (youknowriad) [reply packages/edit-patterns/src/components/index.js]: about the "reusable-blocks" package. I'm not a big fan of that one much either but I think it's probably "ok", let's just document that this is a block-editor specific component.

Comment (glendaviesnz) [reply packages/edit-patterns/src/components/index.js]: Thanks for the additional context. I have updated the [package readme](https://github.com/WordPress/gutenberg/pull/53161/files#diff-87865f128cc1dc55d61ae27ec3600a115663235a8aa57d6ef90bc7bbaec6109fR15). There is no urgency for this merge so will keep it open for a few days, so get back to me if you have any additional thoughts about a better location for these components.

File: packages/patterns/README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+```bash
+npm install @wordpress/patterns --save
+```
+
+_This package assumes that your code will run in an **ES2015+** environment. If you're using an environment that has limited or no support for such language features and APIs, you should include [the polyfill shipped in `@wordpress/babel-preset-default`](https://github.com/WordPress/gutenberg/tree/HEAD/packages/babel-preset-default#polyfill) in your code._
+
+## How it works
+
+This package is used internally by the Gutenberg project to manage the creation and editing of user patterns using the `wp_block` CPT in the context of the block editor. The likes of the `PatternsMenuItems` component expect to be rendered within a `BlockEditorProvider` in order to work.
```
Comment (aaronrobertshaw) [inline packages/patterns/README.md]: Is it worth raising the prominence of this via `warning` or `note` notices, or making it the first part of the readme? I'm thinking of how the [interactivity package conveys](https://github.com/WordPress/gutenberg/blob/trunk/packages/interactivity/README.md?plain=1#L1-L11) similar info.

Comment (glendaviesnz) [reply packages/patterns/README.md]: Have moved it to a note at the top of the readme.

File: packages/patterns/package.json [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	"keywords": [
+		"wordpress",
+		"gutenberg",
+		"patterns"
+	],
+	"homepage": "https://github.com/WordPress/gutenberg/tree/HEAD/packages/patterns/README.md",
+	"repository": {
+		"type": "git",
+		"url": "https://github.com/WordPress/gutenberg.git",
+		"directory": "packages/editor"
```
Comment (andrewserong) [inline packages/patterns/package.json]: Should this be `patterns`?

```suggestion
		"directory": "packages/patterns"
```

Comment (glendaviesnz) [reply packages/patterns/package.json]: ah, well spotted - thanks

**Comment by glendaviesnz** [comment]
This is rebased with node update now, and ready for a final test and ✅  when someone has time, please.

**Review by kevin940726 — APPROVED** [review]
LGTM 👍

**Review by andrewserong — APPROVED** [review]
Great PR!

✅ Implementation matches all the feedback so far
✅ Creating synced and unsynced patterns, detaching, and managing patterns works in post editor
✅ Creating synced and unsynced patterns, detaching, and managing patterns works in site editor block editor
✅ Creating synced and unsynced patterns works correctly in the site editor's browse mode patterns management screen

Thanks for the nicely organised scope of the PR, too, it made it straightforward to review even though there's lots of files 🎉 

LGTM! ✨
